### DISCUSSION TITLE
Best strategy for making React component library Tree Shakable?

### DISCUSSION THREAD

**Discussion by jordanlatimer** [webpack/webpack#12579] [Q&A] [answered]
I'm making a component library for React, but I can't seem to tree shake it when I use it with my other projects. I've read the documentation, issues on this repo, and multiple StackOverflow questions, but I can't seem to find a solution and must be missing something. Note that I'm not bundling the library with Webpack, but I am bundling the projects that use the library with Webpack.

My components all use named exports:
```
//component1.js
export const Component1 = () => {
  return(
    //...
  )
}
```
```
//component2.js
export const Component2 = () => {
  return(
    //...
  )
}
```
Then, in an index.js file:
```
//index.js
export { Component1 } from "./component1.js"
export { Component2 } from "./component2.js"
```
When only Component1 is used, Component 2 is still bundled. I tried putting `sideEffects: false` in my package.json, but that didn't work, presumably because at least one export from this module is used, as the [docs say](https://webpack.js.org/guides/tree-shaking/#clarifying-tree-shaking-and-sideeffects) that "If no direct export from a module flagged with no-sideEffects is used, the bundler can skip evaluating the module for side effects.".

I would like to have a single index file export all my components, so I decided to try out the `/*#__PURE__*/` annotation. I imported all the components and then used an IIFE to allow them to be marked as pure:
```
//index.js
import { Component1 as C1 } from "./component1.js";
export const Component1 = /*#__PURE__*/ (() => C1)();
import { Component1 as C2 } from "./component1.js";
export const Component2 = /*#__PURE__*/ (() => C2)();
```
But this didn't work either. Perhaps I'm using the `/*#__PURE__*/` annotation wrong?

How do people get component libraries to be tree shakeable? Do I need to stop using an index file and instead follow the approach of a library like MaterialUI, where each component is [imported separately](https://material-ui.com/guides/minimizing-bundle-size/#development-environment)? Which if I do need to do that, fine, but I'd rather not if I can.

**Comment by sokra** [comment, marked as answer]
> //component1.js
> export const Component1 = () => {
>   return(
>     //...
>   )
> }
> 
> //component2.js
> export const Component2 = () => {
>   return(
>     //...
>   )
> }
> 
> Then, in an index.js file:
> //index.js
> export { Component1 } from "./component1.js"
> export { Component2 } from "./component2.js"

Plus `sideEffects: false`

That should work this way.

**Reply by jordanlatimer** [reply]
That's what I thought, but for some reason it's not working. Perhaps it's an issue with the project I have consuming this library?

Should `sideEffects: false` also be used in the consuming project's project.json?

**Reply by jordanlatimer** [reply]
Testing it with `@material-ui/core`, it appears that it won't tree shake that library either. Perhaps it IS something with my consuming project...

I'm testing the library using an npm symlink, would that cause anything?

**Reply by jordanlatimer** [reply]
Nevermind, I figured it out. I was using ts-loader and it was converting it to a non-es5 format.

**Reply by nyngwang** [reply]
@jordanjlatimer: I guess you wanted to say non-**es6** format.
